### PR TITLE
Add a new experimental API to set additional image operands

### PR DESCRIPTION
**Pull request by oisyn**
This PR adds a new experimental API to set additional image operands when sampling textures. Currently, it supports `bias`, `lod`, `grad` and `sample_index` operands.

Usage:
```rust
use spirv_std::spirv;
use spirv_std::{arch, image::sample_with, image::ImageWithMethods, Image, Sampler};

#[spirv(fragment)]
pub fn main(
    #[spirv(descriptor_set = 0, binding = 0)] sampler: &Sampler,
    #[spirv(descriptor_set = 0, binding = 1)] image1: &Image!(2D, type=f32, sampled, multisampled),
    #[spirv(descriptor_set = 0, binding = 2)] image2: &Image!(2D, type=f32, sampled),
    output: &mut glam::Vec4,
) {
    let t1 = image1.fetch_with(glam::IVec2::new(0, 0), sample_with::sample_index(1));
    let t2 = image2.sample_with(*sampler, glam::Vec2::new(0.5, 0.5), sample_with::bias(1.0));
    let t3 = image2.sample_with(*sampler, glam::Vec2::new(0.5, 0.5), sample_with::lod(2.0));
    let t4 = image2.sample_with(
        *sampler,
        glam::Vec2::new(0.5, 0.5),
        sample_with::grad(glam::Vec2::new(0.5, 0.5), glam::Vec2::new(0.5, 0.5)),
    );
    *output = t1 + t2 + t3 + t4;
}
```

Technically the API supports chaining of multiple operands, e.g. `sample_with::bias(3.0).sample_index(1)`, although none of the currently supported operands can logically be combined.

The sampling methods are part of the `image::ImageWithMethods` trait, and it supports the following methods:
* `fetch_with`
* `gather_with`
* `sample_with`
* `sample_depth_reference_with`
* `sample_with_project_coordinate_with`
* `sample_depth_reference_with_project_coordinate_with`

Future improvements: support more operands, pervent combining mutually exclusive operands, and better error messaging when you use an unsupported operand on a specific sampling function.

### CODE REVIEW

File: crates/spirv-std/macros/src/lib.rs [hunk 1]
```diff
@@ -625,3 +625,132 @@ fn debug_printf_inner(input: DebugPrintfInput) -> TokenStream {
...
+/// instruction ending with a placeholder `$PARAMS` operand. The last parameter must be named
+/// `params`, its type will be rewritten. Relevant generic arguments are added to the function
+/// signature. See `SAMPLE_PARAM_TYPES` for a list of names you cannot use as generic arguments.
+#[proc_macro_attribute]
+#[doc(hidden)]
+pub fn gen_sample_param_permutations(_attr: TokenStream, item: TokenStream) -> TokenStream {
+    let item_fn = syn::parse_macro_input!(item as syn::ItemFn);
+    let mut fns = Vec::new();
+
+    for m in 1..((1 << SAMPLE_PARAM_COUNT) - 1) {
```
Comment (eddyb) [inline crates/spirv-std/macros/src/lib.rs:751]:
```suggestion
    for m in 1..(1 << SAMPLE_PARAM_COUNT) {
```
(or `..=` with the `- 1`)

Comment (oisyn) [reply crates/spirv-std/macros/src/lib.rs:751]: Oh good call. I had it correct when I was still hardcoding those numbers 😄

File: crates/spirv-std/macros/src/lib.rs [hunk 2]
```diff
@@ -625,3 +625,149 @@ fn debug_printf_inner(input: DebugPrintfInput) -> TokenStream {
...
+#[doc(hidden)]
+pub fn gen_sample_param_permutations(_attr: TokenStream, item: TokenStream) -> TokenStream {
+    let item_impl = syn::parse_macro_input!(item as syn::ItemImpl);
+    let mut fns = Vec::new();
+
+    for m in 1..(1 << SAMPLE_PARAM_COUNT) {
+        fns.push(SampleImplRewriter::rewrite(m, &item_impl));
+    }
+
+    println!("{}", quote! { #(#fns)* }.to_string());
```
Comment (oisyn) [inline crates/spirv-std/macros/src/lib.rs:771]: todo: remove debug stuff

File: crates/spirv-std/src/image/sample.rs [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+}
+/// Helper struct that denotes that the type doesn't exist, analog to `Option::None`
+pub struct NoneTy;
+
+/// Helper struct that denotes that the type does exist and is of type T, analog to `Option::Some(T)`
+pub struct SomeTy<T>(T);
+
+/// Helper struct that allows building image operands. Start with a global function that returns this
+/// struct, and then chain additional calls.
+/// Example: `image.sample_with(coords, params::bias(3.0).sample_index(1))`
```
Comment (oisyn) [inline crates/spirv-std/src/image/sample.rs:22]: `params::` is outdated, moved to `sample` mod.

File: crates/spirv-std/macros/src/lib.rs [hunk 4]
```diff
@@ -625,3 +625,149 @@ fn debug_printf_inner(input: DebugPrintfInput) -> TokenStream {
...
+            if altered {
+                self.add_regs(&mut new_t);
+            }
+
+            m.tokens = new_t.into_iter().collect();
+        }
+    }
+}
+
+/// Generates permutations of a sampling function containing inline assembly with an image
```
Comment (oisyn) [inline crates/spirv-std/macros/src/lib.rs:757]: Outdated comment

**Review by eddyb — APPROVED** [review]

**Pull request merged**